File: MultiProcessTest/src/main/java/com/lilong/multiprocesstest/MainActivity.java
package com.lilong.multiprocesstest;

import android.app.Activity;
import android.app.Application;
import android.content.Intent;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

/**
 * -------------------关于多个launcher activity------------------------------------------
 * 所有在manifest中注册时, intentFilter中action写了{@link Intent.ACTION_MAIN}, category中写了{@link Intent.CATEGORY_LAUNCHER}的activity
 * 都会在桌面上出现图标, 也就是桌面上会出现多个图标
 *
 * (1) 在应用被杀死后, 点击图标会进自己
 * (2) 应用在后台, 点击图标会进之前打开的activity
 *
 * 如果卸载掉其中一个, 另一个也会跟着卸载，就是同一个应用，只是有两个入口了
 *
 * -------------------关于多进程--------------------------------------------------------
 * (1) 一个应用可以有多个进程, 进程名也可以互相没关系, 进程中可以运行四大组件和其他代码, 也都有主线程/消息循环
 * (2) 每个进程都有自己的虚拟机
 * (3) 进程之间内存是隔离的, 类加载器也不同, 在不同进程中对同一个变量的修改互不影响, 比如这个{@link #flag}
 * (4) 每个进程在启动时都会构造自己的{@link Application}对象, 并调用其{@link Application#onCreate()}方法
 * (5) 虽然这些{@link Application}对象的{@link Application#hashCode()}的值是一样的, 但它们不是同一个对象(属于不同进程)
 *     其他的组件比如主线程, {@link Looper}等也是同样的情况
 * (6) {@link SecondActivity}配置在"global.process"中, 进程名以小写字母开头, 表示这是个全局进程, 其它同签名同sharedUID的应用的组件, 也可以跑在这个进程里
 * (7) {@link ThirdActivity}配置在":private.process"中, 进程名以冒号开头, 表示这是个私有进程, 其他应用的组件不可以跑在这个进程里
 *
 * */
public class MainActivity extends Activity {

    public static final String TAG = "MTest";

    private MenuItem menuItemJumpToSecondActivity;
    private MenuItem menuItemJumpToThirdActivity;

    public static String flag = "MainActivity";

    private static final int CODE_SECOND_ACTIVITY = 1;
    private static final int CODE_THIRD_ACTIVITY = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setTitle("MainActivity");
        Log.i(TAG, "currentThread = " + Thread.currentThread() + ", looper = " + Looper.myLooper());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.actions, menu);
        menuItemJumpToSecondActivity = menu.findItem(R.id.menuItemJumpToSecondActivity);
        menuItemJumpToThirdActivity = menu.findItem(R.id.menuItemJumpToThirdActivity);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item == menuItemJumpToSecondActivity){
            Intent intent = new Intent(MainActivity.this, SecondActivity.class);
            startActivityForResult(intent, CODE_SECOND_ACTIVITY);
        } else if(item == menuItemJumpToThirdActivity){
            Intent intent = new Intent(MainActivity.this, ThirdActivity.class);
            startActivityForResult(intent, CODE_THIRD_ACTIVITY);
        }
        return true;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode){
            case CODE_SECOND_ACTIVITY:
                Log.i(TAG, "returned from SecondActivity, flag = " + flag);
                break;
            case CODE_THIRD_ACTIVITY:
                Log.i(TAG, "returned from ThirdActivity, flag = " + flag);
                break;
        }
    }
}

File: SerializableTest/src/main/java/com/lilong/serializabletest/MainActivity.kt
package com.lilong.serializabletest

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.onClick
import java.io.*

/**
 * serializable和parcelable的区别：
 *
 * serializable:
 * (1) java库中的功能
 * (2) 使用简单，代码少，只要让类实现Serializable接口即可，几乎全自动
 * (3) 序列化过程中通过反射来解析，速度比parcelable慢几乎十倍
 * (4) 占用内存较多，会生成大量临时对象，内存抖动厉害
 * (5) 可以通过ObjectOutputStream将序列化的数据写入磁盘/通过网络传输
 *
 * parcelable:
 * (1) android独有的功能
 * (2) 使用复杂，代码多，需要指定对象与Parcel之间互相转换的过程（需要实现writeToParcel和CREATOR中的createFromParcel)，几乎全手动
 * (3) 序列化过程会依照(2)中用户指定的过程来进行，速度很快，效率高
 * (4) 占内存少
 * (5) 虽然技术上可以，但不应该将parcel的数据（通过parcel.marshall方法可获得其字节数据）写入磁盘或通过网络传输，因为parcel是根据机器和环境深度优化过的，没有通用性，换机器后很可能无法读取
 * */
class MainActivity : Activity() {

    val objFile: File = File(MainApplication.Companion.getInstance()?.cacheDir?.absolutePath + File.separator + "haha")
    var menuItemJumpToActivityInAnotherProcess: MenuItem? = null;

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.option, menu)
        menuItemJumpToActivityInAnotherProcess = menu?.findItem(R.id.jumpToActivityInAnotherProcess)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item == menuItemJumpToActivityInAnotherProcess) {
            var intent: Intent = Intent(this, SecondActivity::class.java)
            startActivity(intent)
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setTitle("MainActivity")
        var student: Student = Student("xiaoming", 12)
        tvBeforeSerialize.text = student.toString()
        btnSerialize.onClick {
            var objectOutputStream: ObjectOutputStream = ObjectOutputStream(FileOutputStream(objFile))
            objectOutputStream.writeObject(student)
            objectOutputStream.close()
        }

        /**
         * Serializable接口的实例，可以实现序列化和反序列化，通过ObjectInputStream的readObject和ObjectOutputStream的writeObject方法
         * 可以将对象的数据（属性的值）都存储和还原
         * */
        btnDeSerialize.onClick {
            try {
                var objectInputStream: ObjectInputStream = ObjectInputStream(FileInputStream(objFile))
                var student: Student = objectInputStream.readObject() as Student
                objectInputStream.close()
                tvAfterDeSerialize.text = student.toString()
            } catch (e: Exception) {
            }
        }
    }
}
